### ISSUE TITLE
backport fix for Const propagation of uninitialized structs to `v1.1`.

### ISSUE THREAD

**Issue by rai-nhdaly** [JuliaLang/julia#31418]
Opening an issue about this so we don't forget.

Today we noticed our code incorrectly const-propagating an unitialized tuple on `julia 1.1`, which it wasn't doing on `julia 1.0`.

This is fixed on `master` by the following lines, introduced in https://github.com/JuliaLang/julia/commit/52bafeb981bac548afd2264edb518d8d86944dca:
https://github.com/JuliaLang/julia/blob/f611b46e72c788c016933033dfadee6a9e138d47/base/compiler/abstractinterpretation.jl#L931-L937

We discussed backporting this fix to a patch release for `v1.1`. Does that still seem doable? 😊Would we backport just those lines, or the whole commit https://github.com/JuliaLang/julia/commit/52bafeb981bac548afd2264edb518d8d86944dca?

Thanks! :)

**Comment by rai-nhdaly** [comment]
As-is, we are still unable to use julia `v1.1` because LLVM takes 3 hours trying to compile a single function when trying to const-propagate an uninitialized struct (a `Nullables.Nullable{LargeStruct}` that starts as `nothing`). This bug was introduced in `v1.1`, and doesn't happen on `v1.0.3`.

Would it be possible to back-port all of the commit 52bafeb? If not, is there an easy way to fix this problem for backport?
Thanks

**Comment by rai-nhdaly** [comment]
😁@Keno, I see that @vchuravy assigned you to this. Haha sorry and thanks. :)

Let me know if I can help you with this at all. This is currently making `v1.1` unusable for us without running on `-O1`, but that of course makes our benchmarking less reliable.

Thanks!

**Comment by Keno** [comment]
I don't think it's very hard. It should literally just be about adding that `if` statement (the `fieldcount(t) == length(ats)`, but of course on 1.1, I don't think you need to actually construct the `ats`, you can just count it)

**Comment by rai-nhdaly** [comment]
Cool, good to hear.

So for `v1.1`, you think the diff would basically just be something like this?:
```diff
diff --git a/base/compiler/abstractinterpretation.jl b/base/compiler/abstractinterpretation.jl
index b01ad3a271..e493b4a1b9 100644
--- a/base/compiler/abstractinterpretation.jl
+++ b/base/compiler/abstractinterpretation.jl
@@ -910,7 +910,7 @@ function abstract_eval(@nospecialize(e), vtypes::VarTable, sv::InferenceState)
                     isconst = false
                 end
             end
-            if isconst
+            if isconst && fieldcount(t) == length(length(e.args) - 1)
                 t = Const(ccall(:jl_new_structv, Any, (Any, Ptr{Cvoid}, UInt32), t, args, length(args)))
             end
         end
```

**Comment by Keno** [comment]
yes, something like that. If that works and solves your issue, submit it as a PR :)

**Comment by rai-nhdaly** [comment]
K, i'll look into it. Thanks for the support, Keno. :)

**Comment by rai-nhdaly** [comment]
Thanks Keno, that did indeed fix our issue. I've opened a PR here:
https://github.com/JuliaLang/julia/pull/31699

**Comment by rai-nhdaly** [comment]
This has been addressed in #31699. Closing.
